### ISSUE TITLE
Batch size inconsistency when averaging loss 

### ISSUE THREAD

**Issue by nluedema** [uma-pi1/kge#172]
I had a look at the different training jobs and noticed an inconsistency in how the loss is averaged between `TrainingJob1vsAll` and `TrainingJobKvsAll`/`TrainingJobNegativeSampling`. All three jobs divide the loss by `batch_size`, but in `TrainingJob1vsAll` the `batch_size` is based on an object which was indexed by `subbatch_slice` while in `TrainingJobKvsAll`/`TrainingJobNegativeSampling` the `batch_size` is based on an object which was not indexed by `subbatch_slice`.

Here are the relevant parts in the code:
```
#TrainingJob1vsAll
triples = batch["triples"][subbatch_slice].to(self.device)
batch_size = len(triples)
```
```
# TrainingJobNegativeSampling
triples = batch["triples"][subbatch_slice]
batch_negative_samples = batch["negative_samples"]
batch_size = len(batch["triples"])
```
```
#TrainingJobKvsAll
queries_subbatch = batch["queries"][subbatch_slice].to(self.device)
batch_size = len(batch["queries"])
```

I was looking at this, because I want to create unequally sized sub batches of the sub batches, based on the relation index of the triples (i.e. subsub batches). Intuitively, I would have divided the loss of a subsub batch by the size of that batch. However, the approach in `TrainingJobKvsAll`/`TrainingJobNegativeSampling` suggests I should use the overall batch size

**Comment by samuelbroscheit** [comment]
I agree that the demominator in TrainingJob1vsAll._process_subbatch() should be the true batch_size not the subbatch size. Or are we missing something @rgemulla ?

**Comment by samuelbroscheit** [comment]
The challenge is that when we fix this, how we can take care of users that have hyperparameter settings that were obtained before the fix. Changing this will affect the learning rate ect.

**Comment by rgemulla** [comment]
Yes, we should fix this and add a test that checks equality of losses with and without the usage of subbatches.

Subbatches are a relatively new feature and the bug only matters when subbatches are used. So the hope is that few people would be affected.

**Comment by rgemulla** [comment]
We should directly use ```result.size``` everywhere, then it's consistent. I'll do a quick PR.

**Comment by rgemulla** [comment]
@nluedema Please have a look at #173 and see if you agree
